src/invoke-lambda/spawn.js

```diff
@@ -54,44 +54,45 @@ module.exports = function spawnChild (params, callback) {
   // Ensure we don't have dangling processes due to open connections, etc.
   function maybeShutdown (event) {
     update.debug.status(`${functionPath} (pid ${pid}) shutting down (via ${event} event)`)
-    if (isRunning(pid) && !murderInProgress && !closed) {
-      update.debug.status(`${functionPath} (pid ${pid}) is still running, terminating now...`)
-      murderInProgress = true
-      let code = 0
-      if (error) {
-        update.error(`${functionPath} (pid ${pid}) caught hanging execution with an error, attempting to exit 1`)
-        code = 1
-      }
-      if (!isInLambda) {
-        kill(pid, SIG, () => {
-          murderInProgress = false
-          update.debug.status(`${functionPath} (pid ${pid}) successfully terminated`)
-          if (!closed) done(code)
-        })
-      }
-      else {
-        // tree-kill relies on *nix `ps`, which Lambda doesn't have – but it does have /proc
-        // Node process.kill() + Lambda Linux /proc/<pid>/task/<tid> is mysterious, so this may not be the best or proper approach
-        try {
-          let tasks = readdirSync(`/proc/${pid}/task`)
-          tasks.forEach(tid => {
-            try { process.kill(tid) }
-            catch (err) {
-              // Task may have ended naturally or been killed by killing child.pid, I guess we don't really know
-              update.debug.status(`${functionPath} (pid ${pid}) did not kill task (tid ${tid})`)
-            }
-          })
-          update.debug.status(`${functionPath} (pid ${pid}) (possibly maybe) successfully terminated inside Lambda`)
+    if (!isRunning(pid) || murderInProgress || closed) {
+      return update.debug.status(`${functionPath} (pid ${pid}) is not running (termination in progress: ${murderInProgress}; process closed: ${closed}`)
+    }
+    update.debug.status(`${functionPath} (pid ${pid}) is still running, terminating now...`)
+    murderInProgress = true
+    let code = 0
+    if (error) {
+      update.error(`${functionPath} (pid ${pid}) caught hanging execution with an error, attempting to exit 1`)
+      code = 1
+    }
+    if (!isInLambda) {
+      kill(pid, SIG, () => {
+        murderInProgress = false
+        update.debug.status(`${functionPath} (pid ${pid}) successfully terminated`)
+        if (!closed) {
+          done(code)
         }
+      })
+      return
```

Comment:
That `kill` call is asynchronous, returning there might cause a race condition / weird side effects, no?

Reply:
If we return before `kill` has completed, `murderInProgress` may not have been set, and `done` may not have been called; if in that time `maybeShutdown` is called again, the state of the predicates will not be accurate. I don't see the necessity of returning here, but I can def see weird cases where it can cause issues.